crates/aptos-dkg/src/range_proofs/dekart_univariate_v2.rs

```diff
@@ -0,0 +1,809 @@
+        // Step 4b
+        fiat_shamir::append_f_j_commitments::<E>(fs_t, &Cj);
+
+        // Step 6
+        let (beta, betas) = fiat_shamir::get_beta_challenges::<E>(fs_t, ell);
+
+        let hat_f_evals: Vec<E::ScalarField> = {
+            let mut v = Vec::with_capacity(num_omegas);
+            v.push(r);
+            v.extend_from_slice(values);
+            v.resize(num_omegas, E::ScalarField::ZERO);
+            v
+        };
+
+        let hat_f_coeffs = eval_dom.ifft(&hat_f_evals);
+        debug_assert_eq!(hat_f_coeffs.len(), pk.max_n + 1);
+
+        let diff_hat_f_evals: Vec<E::ScalarField> = {
+            let coeffs = polynomials::differentiate(&hat_f_coeffs);
+            eval_dom.fft(&coeffs)
+        };
+
+        let f_j_coeffs: Vec<Vec<E::ScalarField>> = (0..ell)
+            .map(|j| {
+                let mut f_j = f_js_evals[j].clone();
+                debug_assert_eq!(f_j.len(), pk.max_n + 1);
+                eval_dom.ifft_in_place(&mut f_j);
+                debug_assert_eq!(f_j.len(), pk.max_n + 1);
+                f_j
+            })
+            .collect();
+
+        let diff_f_js_evals: Vec<Vec<E::ScalarField>> = f_js_evals
+            .iter()
+            .map(|f_j_eval| {
+                let mut coeffs = f_j_eval.clone();
+                eval_dom.ifft_in_place(&mut coeffs); // Convert to coefficients
+                let diff_coeffs = polynomials::differentiate(&coeffs); // Differentiate
+                eval_dom.fft(&diff_coeffs) // Convert back to evaluations
+            })
+            .collect();
+
+        let h_evals: Vec<E::ScalarField> = {
+            let mut h = Vec::with_capacity(num_omegas);
+
+            let first_h_eval = {
+                let mut pow2 = E::ScalarField::ONE;
+                let mut sum_pow2_rs = E::ScalarField::ZERO;
+                for r_j in &rs {
+                    sum_pow2_rs += pow2 * r_j;
+                    pow2 = pow2.double();
+                }
+
+                let sum_betas_term: E::ScalarField = betas
+                    .iter()
+                    .zip(&rs)
+                    .map(|(beta_j, r_j)| *beta_j * r_j * (*r_j - E::ScalarField::ONE))
+                    .sum();
+
+                let numerator = beta * (r - sum_pow2_rs) + sum_betas_term;
+                numerator / E::ScalarField::from(num_omegas as u64) // TODO!!!!!!!
+            };
+            h.push(first_h_eval);
+
+            for i in 1..num_omegas {
+                // First term: beta * diff_hat_f_evals[i]
+                let mut val = diff_hat_f_evals[i];
+
+                // Second term: -beta * sum_j 2^j * f_j_evals[j][i]
+                let sum1: E::ScalarField = diff_f_js_evals
+                    .iter()
+                    .enumerate()
+                    .map(|(j, diff_f_j)| E::ScalarField::from(1u64 << j) * diff_f_j[i])
+                    .sum();
+                val = (val - sum1) * beta;
+
+                // Third term: sum_j betas[j] * diff_f_j * (2*f_j - 1)
```

Comment:
I think you are invoking finite field arithmetic unnecessarily here: you can just look at the jth bit of the ith value and output either +1 or -1 based on it. But I am not sure if the finite field arithmetic costs you a lot here (e.g., would 0 - 1 perform a modular reduction?)